mlir/include/mlir/Dialect/Vector/IR/VectorOps.td

```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
   }];
 }
 
+class ResultIsHalfSourceVectorType<string result> : TypesMatchWith<
+  "the trailing dimension of the results is half the width of source trailing dimension",
+  "source", result,
+  [{
+    [&]() -> ::mlir::VectorType {
+      auto vectorType = ::llvm::cast<mlir::VectorType>($_self);
+      ::mlir::VectorType::Builder builder(vectorType);
+      auto lastDim = vectorType.getRank() - 1;
+      auto newDimSize = vectorType.getDimSize(lastDim) / 2;;
+      if (newDimSize <= 0)
+         return vectorType; // (invalid input type)
+      return builder.setDim(lastDim, newDimSize);
+    }()
+  }]
+>;
+
+class SourceVectorEvenElementCount : PredOpTrait<"the trailing dimension of the source vector has an even number of elements",
+    CPred<[{
+      [&](){
+        auto srcVec = getSourceVectorType();
+        return srcVec.getDimSize(srcVec.getRank() - 1) % 2 == 0;
+      }()
+    }]>
+>;
+
+def Vector_DeinterleaveOp :
+  Vector_Op<"deinterleave", [Pure,
+    SourceVectorEvenElementCount<>,
+    ResultIsHalfSourceVectorType<"res1">,
+    ResultIsHalfSourceVectorType<"res2">,
+    AllTypesMatch<["res1", "res2"]>
+    ]> {
+      let summary = "constructs two vectors by deinterleaving an input vector";
+      let description = [{
+        The deinterleave operation constructs two vectors from a single input
+        vector. The first result vector contains the elements from even indexes
+        of the input, and the second contains elements from odd indexes. This is
+        the inverse of a 'vector.interleave' operation.
+
+        Each output's trailing dimension is half of the size of the input
+        vector's trailing dimension. This operation requires the input vector
+        to have a rank > 0 and an even number of elements in its trailing
+        dimension.
+
+        The operation supports scalable vectors.
+
+        Example:
+        ```mlir
+        %0, %1 = vector.deinterleave %a
+                   : vector<[4]xi32> into vector<[2]xi32>, vector<[2]xi32>
+        %2, %3 = vector.deinterleave %b
+                   : vector<8xi8> into vector<4xi8>, vector<4xi8>
+        %4, %5 = vector.deinterleave %c
+                   : vector<2x8xf32> into vector<2x4xf32>, vector<2x4xf32>
+        %6, %7 = vector.deinterleave %d
+                   : vector<2x4x[6]xf64> into vector<2x4x[3]xf64>, vector<2x4x[3]xf64>
+        ```
+      }];
+
+      let arguments = (ins AnyVector:$source);
+      let results = (outs AnyVector:$res1, AnyVector:$res2);
+
+      let assemblyFormat = [{
+        $source attr-dict `:` type($source)
+      }];
```

Comment:
The examples and `assemblyFormat` don't match. Also, I think @kuhar's suggestion was just to show one of the result types (since the other is the same). 

e.g.:
```
$source attr-dict `:` type($source) `into` type($res1)
```

Reply:
@MacDue, @kuhar, I've added and `assemblyFormat` that fully describes the operation without being too verbose. An "x2" suffix on the result describes the quantity of that type quite simply and should be easily readable too.

So the resulting IR should look like this:
`%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>x2`

Reply:
That looks good to me :+1:

Reply:
I'm not a fan of that assembly format, I think consistency is much more important and would rather have the verbose form.

Reply:
The verbose form is not consistent with `vector.interleave` (which only shows the source type) -- (similar to `vectpr.shuffle`)

Reply:
I meant consistency in general not in relation to interleave.

Reply:
Should it be consistent to interleave or the other MLIR operation formats?

Reply:
> @MacDue, @kuhar, I've added and `assemblyFormat` that fully describes the operation without being too verbose. An "x2" suffix on the result describes the quantity of that type quite simply and should be easily readable too.
> 
> So the resulting IR should look like this: `%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>x2`

I'd prefer just `%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>` for readability and consistency with most of core mlir dialects.
Separately, I think we should also have the result type on `vector.interleave` (I wasn't pinged on the initial PR IIRC, and I didn't catch it before it landed).